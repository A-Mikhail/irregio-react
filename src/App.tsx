import * as React from 'react';
import styled from './theme/styled-components';

const Title = styled.h1`
  color: ${props => props.color};
`;

class App extends React.Component {
  public render() {
    return <Title color={'#7FDBFF'}>Typescript with styled-components</Title>;
  }
}

export default App;
