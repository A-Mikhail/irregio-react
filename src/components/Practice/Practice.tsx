import * as React from "react";
import styled from "../../theme/styled-components";

import Congratulations from "../Congratulations/Congratulations";

interface IProps {
  progress?: number;
  infinitive: string;
  pastTense?: string;
  presentPerfect?: string;
  translationRus?: string;
  translationEn?: string;
  frequency: string;
  showTranslation: boolean;
  isCompleted: boolean;
  reset?: () => void;
  hintsTaken: number;
  filter?: string;
}

const StyledPractice = styled.div`
  display: flex;
  max-width: 400px;
  margin: auto;
  flex-direction: column;
  text-align: center;
  > div {
    padding: 2px;
    color: white;
  }
`;
const StyledInfinitive = styled.div`
  font-size: 3rem;
`;

const StyledTranslation = styled.div`
  position: relative;
  color: #f0f0f0;
  opacity: 0.4;
  font-size: 2rem;
  top: -15px;
`;

class Learn extends React.Component<IProps> {
  public componentDidMount() {
    if (this.props.reset) {
      this.props.reset();
    }
  }

  public render() {
    const {
      infinitive,
      translationRus,
      translationEn,
      showTranslation,
      isCompleted,
      hintsTaken,
      filter
    } = this.props;
    return (
      <StyledPractice>
        <StyledInfinitive>{infinitive}</StyledInfinitive>
        {showTranslation ? (
          <StyledTranslation>
            {translationEn}/{translationRus}
          </StyledTranslation>
        ) : null}
        {isCompleted ? (
          <Congratulations hintsTaken={hintsTaken} filter={filter} />
        ) : null}
        {this.props.children}
      </StyledPractice>
    );
  }
}

export default Learn;
