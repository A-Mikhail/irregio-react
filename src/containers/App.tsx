import * as React from "react";
import styled from "../theme/styled-components";
import de from "./data";

import ProgressBar from "../components/ProgressBar/ProgressBar";
import Button from "../ui/button/button";

interface IState {
  data?: object;
  progress: number;
  fractionCompleted: number;
  hintsTaken: number;
  currentPerfectForm?: string;
  currentPastForm?: string;
}

const Title = styled.h1`
  color: ${props => props.color};
`;

class App extends React.Component<{}, IState> {
  public refOne: HTMLInputElement;
  public refTwo: HTMLInputElement;

  public state = {
    data: de,
    progress: 0,
    fractionCompleted: 0,
    hintsTaken: 0,
    currentPastForm: "",
    currentPerfectForm: ""
  };

  public componentDidMount() {
    document.addEventListener("keypress", this.handleRightArrow);
    document.addEventListener("keypress", this.handleLeftArrow);
  }
  public componentDidUpdate() {
    if (
      this.state.currentPastForm ===
      this.state.data.data[this.state.progress].pastTense
    ) {
      this.refTwo.focus();
    }
    if (
      this.state.currentPerfectForm ===
      this.state.data.data[this.state.progress].presentPerfect
    ) {
      this.refOne.focus();
      this.setState({
        currentPastForm: "",
        currentPerfectForm: "",
        progress: this.state.progress + 1
      });
    }
  }

  public handleButton = (type: string) => {
    if (
      type === "incr" &&
      this.state.progress !== this.state.data.data.length - 1
    ) {
      this.setState({
        ...this.state,
        progress: this.state.progress + 1,
        fractionCompleted:
          this.state.fractionCompleted + 100 / this.state.data.data.length
      });
    } else if (type === "decr" && this.state.progress === 1) {
      this.setState({
        ...this.state,
        progress: 0,
        fractionCompleted: 0
      });
    } else if (type === "decr" && this.state.progress !== 0) {
      {
        this.setState({
          ...this.state,
          progress: this.state.progress - 1,
          fractionCompleted:
            this.state.fractionCompleted - 100 / this.state.data.data.length
        });
      }
    }
  };

  public handlePastForm = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ currentPastForm: event.target.value });
  };
  public handlePerfectForm = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ currentPerfectForm: event.target.value });
  };

  public handleRightArrow = (event: any) => {
    if (event.keyCode === 39) {
      this.setState({
        ...this.state,
        progress: this.state.progress + 1,
        fractionCompleted:
          this.state.fractionCompleted + 100 / this.state.data.data.length
      });
    }
  };
  public handleLeftArrow = (event: any) => {
    if (event.keyCode === 37 && this.state.progress !== 0) {
      this.setState({
        ...this.state,
        progress: this.state.progress - 1,
        fractionCompleted:
          this.state.fractionCompleted - 100 / this.state.data.data.length
      });
    }
  };

  public render() {
    console.log(this.state.progress);
    const { data } = this.state.data;
    return (
      <React.Fragment>
        <Title color={"#7FDBFF"}>Irreg.io</Title>
        <div>Progress: {this.state.progress}</div>
        <div>Infinitive: {data[this.state.progress].infinitive}</div>
        <div>Past form: {data[this.state.progress].pastTense}</div>
        <div>Perfect form: {data[this.state.progress].presentPerfect}</div>
        <div>
          Russian translation: {data[this.state.progress].translationRus}
        </div>
        <div>
          English translation: {data[this.state.progress].translationEn}
        </div>
        <div>Frequency: {data[this.state.progress].frequency}</div>
        <div>Fraction completed: {this.state.fractionCompleted}</div>
        <Button label="Next" clicked={() => this.handleButton("incr")} />
        <Button label="Previous" clicked={() => this.handleButton("decr")} />
        <input
          ref={refOne => (this.refOne = refOne as HTMLInputElement)}
          onChange={this.handlePastForm}
          type="text"
          value={this.state.currentPastForm}
          placeholder="past tense"
        />
        <input
          ref={refTwo => (this.refTwo = refTwo as HTMLInputElement)}
          onChange={this.handlePerfectForm}
          type="text"
          value={this.state.currentPerfectForm}
          placeholder="perfect tense"
        />
        <ProgressBar fractionCompleted={this.state.fractionCompleted} />
      </React.Fragment>
    );
  }
}

export default App;
