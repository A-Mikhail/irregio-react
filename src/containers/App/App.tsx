import * as React from "react";
import styled from "../../theme/styled-components";
import de from "../data/data";

import ProgressBar from "../../components/ProgressBar/ProgressBar";
import Button from "../../ui/button/button";

interface IState {
  verbs?: object;
  filter?: string;
  filteredVerbs?: object;
  progress: number;
  fractionCompleted: number;
  isCompleted: boolean;
  hintsTaken: number;
  currentPerfectForm?: string;
  currentPastForm?: string;
  pastFormHint?: string;
  perfectFormHint?: string;
  showTranslation?: boolean;
}

const Title = styled.h1`
  color: ${props => props.color};
`;

class App extends React.Component<{}, IState> {
  public refOne: HTMLInputElement;
  public refTwo: HTMLInputElement;

  public state = {
    verbs: de,
    filter: "",
    filteredVerbs: de.data,
    progress: 0,
    fractionCompleted: 0,
    hintsTaken: 0,
    currentPastForm: "",
    currentPerfectForm: "",
    pastFormHint: "",
    perfectFormHint: "",
    isCompleted: false,
    showTranslation: true
  };

  public componentDidMount() {
    document.addEventListener("keypress", this.handleRightArrow);
    document.addEventListener("keypress", this.handleLeftArrow);
  }
  public componentDidUpdate() {
    if (
      this.state.currentPastForm ===
      this.state.filteredVerbs[this.state.progress].pastTense
    ) {
      this.refTwo.focus();
    }
    if (
      this.state.currentPerfectForm ===
        this.state.filteredVerbs[this.state.progress].presentPerfect &&
      this.state.progress !== this.state.filteredVerbs.length - 1
    ) {
      this.refOne.focus();
      this.setState({
        currentPastForm: "",
        currentPerfectForm: "",
        progress: this.state.progress + 1,
        fractionCompleted:
          this.state.fractionCompleted + 100 / this.state.filteredVerbs.length,
        pastFormHint: "",
        perfectFormHint: ""
      });
    } else if (
      this.state.currentPerfectForm ===
        this.state.filteredVerbs[this.state.progress].presentPerfect &&
      this.state.progress === this.state.filteredVerbs.length - 1
    ) {
      this.setState({
        currentPastForm: "",
        currentPerfectForm: "",
        fractionCompleted: 100,
        isCompleted: true
      });
    }
  }
  public handlePastForm = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ currentPastForm: event.target.value });
  };
  public handlePerfectForm = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ currentPerfectForm: event.target.value });
  };

  public handleButton = (type: string) => {
    if (
      type === "incr" &&
      this.state.progress === this.state.filteredVerbs.length - 1
    ) {
      this.setState({
        ...this.state,
        progress: this.state.filteredVerbs.length - 1,
        fractionCompleted: 100
      });
    } else if (
      type === "incr" &&
      this.state.progress !== this.state.filteredVerbs.length - 1
    ) {
      this.setState({
        ...this.state,
        progress: this.state.progress + 1,
        fractionCompleted:
          this.state.fractionCompleted + 100 / this.state.filteredVerbs.length
      });
    }
    if (type === "decr" && this.state.progress === 1) {
      this.setState({
        ...this.state,
        progress: 0,
        fractionCompleted: 0
      });
    } else if (type === "decr" && this.state.progress !== 0) {
      this.setState({
        ...this.state,
        progress: this.state.progress - 1,
        fractionCompleted:
          this.state.fractionCompleted - 100 / this.state.filteredVerbs.length
      });
    }
  };

  public handleRightArrow = (event: any) => {
    if (
      event.keyCode === 39 &&
      this.state.progress === this.state.filteredVerbs.length - 1
    ) {
      this.setState({
        ...this.state,
        progress: this.state.filteredVerbs.length - 1,
        fractionCompleted: 100
      });
    } else if (
      event.keyCode === 39 &&
      this.state.progress !== this.state.filteredVerbs.length - 1
    ) {
      this.setState({
        ...this.state,
        progress: this.state.progress + 1,
        fractionCompleted:
          this.state.fractionCompleted + 100 / this.state.filteredVerbs.length
      });
    }
  };
  public handleLeftArrow = (event: any) => {
    if (event.keyCode === 37 && this.state.progress === 1) {
      this.setState({
        ...this.state,
        progress: 0,
        fractionCompleted: 0
      });
    } else if (event.keyCode === 37 && this.state.progress !== 0) {
      this.setState({
        ...this.state,
        progress: this.state.progress - 1,
        fractionCompleted:
          this.state.fractionCompleted - 100 / this.state.filteredVerbs.length
      });
    }
  };
  public handleFilter = (f: string, frequency?: string) => {
    this.setState({
      ...this.state,
      filter: f,
      filteredVerbs:
        f === "all"
          ? de.data
          : de.data.filter(entry => entry.frequency === `${frequency}`),
      progress: 0,
      fractionCompleted: 0
    });
  };
  public handleShuffle = () => {
    const currentVerbs = this.state.filteredVerbs;
    const shuffledVerbs = currentVerbs
      .map(a => [Math.random(), a])
      .sort((a: any, b: any): any => a[0] - b[0])
      .map(a => a[1]);

    this.setState({
      filteredVerbs: shuffledVerbs,
      progress: 0,
      fractionCompleted: 0
    });
  };
  public handleHelp = () => {
    if (
      this.state.pastFormHint !==
      this.state.filteredVerbs[this.state.progress].pastTense
    ) {
      this.refOne.focus();
      this.setState({
        ...this.state,
        currentPastForm: "",
        hintsTaken: this.state.hintsTaken + 1,
        pastFormHint: this.state.filteredVerbs[this.state.progress].pastTense
      });
    } else if (
      this.state.currentPerfectForm !==
      this.state.filteredVerbs[this.state.progress].presentPerfect
    ) {
      this.state.currentPastForm ? this.refTwo.focus() : this.refOne.focus();
      this.setState({
        ...this.state,
        currentPerfectForm: "",
        hintsTaken: this.state.hintsTaken + 1,
        perfectFormHint: this.state.filteredVerbs[this.state.progress]
          .presentPerfect
      });
    }
  };
  public toggleShowTranslation = () => {
    this.setState({
      ...this.state,
      showTranslation: !this.state.showTranslation
    });
  };

  public render() {
    const {
      filteredVerbs,
      progress,
      currentPastForm,
      currentPerfectForm,
      pastFormHint,
      perfectFormHint,
      isCompleted,
      fractionCompleted,
      showTranslation
    } = this.state;
    return (
      <React.Fragment>
        <Title color={"#7FDBFF"}>Irreg.io</Title>
        <div>Progress: {progress}</div>
        <div>Infinitive: {filteredVerbs[progress].infinitive}</div>
        <div>Past form: {filteredVerbs[progress].pastTense}</div>
        <div>Perfect form: {filteredVerbs[progress].presentPerfect}</div>
        {showTranslation ? (
          <div>
            <div>
              Russian translation: {filteredVerbs[progress].translationRus}
            </div>
            <div>
              English translation: {filteredVerbs[progress].translationEn}
            </div>
          </div>
        ) : null}
        <div>Frequency: {filteredVerbs[progress].frequency}</div>
        <div>Fraction completed: {fractionCompleted}</div>
        <div>{isCompleted ? <h2>CONGRATS</h2> : null}</div>
        <ProgressBar fractionCompleted={this.state.fractionCompleted} />
        <input
          ref={refOne => (this.refOne = refOne as HTMLInputElement)}
          onChange={this.handlePastForm}
          type="text"
          value={currentPastForm}
          placeholder={pastFormHint ? pastFormHint : " "}
        />
        <input
          ref={refTwo => (this.refTwo = refTwo as HTMLInputElement)}
          onChange={this.handlePerfectForm}
          type="text"
          value={currentPerfectForm}
          placeholder={perfectFormHint ? perfectFormHint : " "}
        />
        <Button label="Next" clicked={() => this.handleButton("incr")} />
        <Button label="Previous" clicked={() => this.handleButton("decr")} />
        <Button
          label="Hard"
          clicked={() => this.handleFilter("hard", "infrequent")}
        />
        <Button
          label="Easy"
          clicked={() => this.handleFilter("easy", "frequent")}
        />
        <Button label="All" clicked={() => this.handleFilter("all")} />
        <Button label="Shuffle" clicked={() => this.handleShuffle()} />
        <Button label="Help" clicked={() => this.handleHelp()} />
        <Button
          label="Show translation"
          clicked={() => this.toggleShowTranslation()}
        />
      </React.Fragment>
    );
  }
}

export default App;
