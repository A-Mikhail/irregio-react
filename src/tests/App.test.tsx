import App from "../containers/App/App";

import * as React from "react";
import * as ReactDOM from "react-dom";

import { shallow } from "enzyme";

import Practice from "../components/Practice/Practice";

import Navigation from "../ui/navi/navigation";

it("App renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe("Practice component", () => {
  const obj = {
    infinitive: "gehen",
    translationRus: "идти",
    translationEn: "go",
    showTranslation: true,
    frequency: "frequent",
    isCompleted: false
  };
  const wrapper = shallow(
    <Practice
      infinitive={obj.infinitive}
      frequency={obj.frequency}
      translationEn={obj.translationEn}
      showTranslation={obj.showTranslation}
      isCompleted={obj.isCompleted}
    />
  );
  it("renders correct number of children", () => {
    expect(wrapper).toHaveLength(1);
    expect(wrapper.dive().children()).toHaveLength(3);
  });
  it("renders correct text", () => {
    expect(wrapper.childAt(0).text()).toBe(`Infinitive: ${obj.infinitive}`);
    expect(wrapper.childAt(2).text()).toBe(`Frequency: ${obj.frequency}`);
    expect(
      wrapper
        .childAt(1)
        .childAt(1)
        .text()
    ).toBe(`English translation: ${obj.translationEn}`);
  });
});
describe("Navigation component", () => {
  it("Has length of 2", () => {
    const wrapper = shallow(<Navigation />);
    expect(wrapper.children()).toHaveLength(2);
  });
});
