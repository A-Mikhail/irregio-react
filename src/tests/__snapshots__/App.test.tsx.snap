// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component renders without crashing 1`] = `
<div
  className="sc-bwzfXH iMxmHM"
>
  <a
    href="/learn"
    onClick={[Function]}
  >
    Learn
  </a>
  <a
    href="/practice"
    onClick={[Function]}
  >
    Practice
  </a>
</div>
`;

exports[`Title component matches snapshot 1`] = `
<div
  className="sc-gzVnrw ilzDvH"
>
  IRREG.IO
</div>
`;
